vaadin-multi-select-list-mixin.html

```diff
@@ -0,0 +1,148 @@
+-->
+<link rel="import" href="vaadin-list-mixin.html">
+<script>
+  window.Vaadin = window.Vaadin || {};
+
+  /**
+   * A mixin for `nav` elements, facilitating multiple selection of childNodes.
+   *
+   * @polymerMixin
+   */
+  Vaadin.MultiSelectListMixin = superClass => class VaadinMultiSelectListMixin extends Vaadin.ListMixin(superClass) {
+    static get properties() {
+      return {
+        /**
+         * Specifies that multiple options can be selected at once.
+         */
+        multiple: {
+          type: Boolean,
+          value: false,
+          reflectToAttribute: true,
+          observer: '_multipleChanged'
+        },
+
+        /**
+         * Array of indexes of the items selected in the items array
+         */
+        selectedValues: {
+          type: Array,
+          notify: true,
+          value: function() {
+            return [];
+          }
+        }
+      };
+    }
+
+    static get observers() {
+      return [
+        `_enhanceMultipleItems(items, multiple, selected, selectedValues, selectedValues.*)`
+      ];
+    }
+
+    ready() {
+      // Should be attached before click listener in list-mixin
+      this.addEventListener('click', e => this._onMultipleClick(e));
+
+      super.ready();
+      this.addEventListener('keydown', e => this._onMultipleKeydown(e));
+    }
+
+    _enhanceMultipleItems(items, multiple, selected, selectedValues) {
+      if (!items || !multiple) {
+        return;
+      }
+
+      if (selectedValues) {
+        const selectedItems = selectedValues.map(selectedId => items[selectedId]);
+        items.forEach(item => item.selected = selectedItems.indexOf(item) !== -1);
+      }
+
+      const lastSelectedItem = this.selectedValues.slice(-1);
+      if (lastSelectedItem && !lastSelectedItem.disabled) {
+        this._scrollToItem(lastSelectedItem);
+      }
+    }
+
+    _onMultipleClick(event) {
+      const item = this._filterItems(event.composedPath())[0];
+      const idx = item && !item.disabled ? this.items.indexOf(item) : -1;
+      if (idx < 0 || !this.multiple) {
+        return;
+      }
+
+      event.preventDefault();
+      if (event.shiftKey) {
+        const lastSelected = this.selectedValues.slice(-1);
+        if (lastSelected !== undefined && lastSelected === idx) {
+          return;
+        }
+        if (this.selectedValues.length) {
+          this.selectedValues = this._getAvailableIndexes(
+            Math.min(lastSelected, idx),
+            Math.max(lastSelected, idx),
+            item => !item.disabled
+          );
+        } else {
+          this.selectedValues = [idx];
+        }
+      } else if (event.metaKey || event.ctrlKey) {
+        this.__updateSelectedValues(idx);
+      } else {
+        this.selectedValues = [idx];
+      }
+    }
+
+    __updateSelectedValues(idx) {
+      if (this.selectedValues.indexOf(idx) !== -1) {
+        this.splice('selectedValues', this.selectedValues.indexOf(idx), 1);
```

Comment:
This causes two `selected-values-changed` events to be fired. Same for below. (So whenever clicking with ctrl)